File: algorithms/0121-best-time-to-buy-and-sell-stock/solution.py
# Best Time to Buy and Sell Stock — track min buy so far; update max profit at each day.
# Time O(n), Space O(1).

class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        min_price = float("inf")
        best = 0
        for price in prices:
            if price < min_price:
                min_price = price
            else:
                best = max(best, price - min_price)
        return best

File: algorithms/0121-best-time-to-buy-and-sell-stock/Solution.java
// Best Time to Buy and Sell Stock - track min buy so far; update max profit at each day.
// Time O(n), Space O(1).

class Solution {
    public int maxProfit(int[] prices) {
        int minPrice = Integer.MAX_VALUE;
        int best = 0;
        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price;
            } else {
                best = Math.max(best, price - minPrice);
            }
        }
        return best;
    }
}
